packages/studio-base/src/components/MessagePipeline/subscriptions.ts

```
@@ -19,6 +18,66 @@ export function makeSubscriptionMemoizer(): (val: SubscribePayload) => Subscribe
   return moize((val: SubscribePayload) => val, { isDeepEqual: true, maxSize: Infinity });
 }
 
+/**
+ * Merge two SubscribePayloads, using either all of the fields or the union of
+ * the specific fields requested.
+ */
+function mergeSubscription(
+  a: Immutable<SubscribePayload>,
+  b: Immutable<SubscribePayload>,
+): Immutable<SubscribePayload> {
+  const isAllFields = a.fields == undefined || b.fields == undefined;
+  const fields = R.pipe(
+    R.chain((payload: Immutable<SubscribePayload>): readonly string[] => payload.fields ?? []),
+    R.map((v) => v.trim()),
+    R.filter((v: string) => v.length > 0),
+    R.uniq,
+  )([a, b]);
+
+  return {
+    ...a,
+    ...(fields.length > 0 && !isAllFields ? { fields } : {}),
+  };
+}
+
+/**
+ * Merge subscriptions that subscribe to the same topic, paying attention to
+ * the fields they need. This ignores `preloadType`.
+ */
+function denormalizeSubscriptions(
+  subscriptions: Immutable<SubscribePayload[]>,
+): Immutable<SubscribePayload[]> {
+  return R.pipe(
+    R.groupBy((v: Immutable<SubscribePayload>) => v.topic),
+    R.values,
+    // Filter out any set of payloads that contains _only_ empty `fields`
+    R.filter((payloads: Immutable<SubscribePayload[]> | undefined) => {
+      // Handle this later
+      if (payloads == undefined) {
+        return true;
+      }
+
+      return !R.all(
+        (v: Immutable<SubscribePayload>) => v.fields != undefined && v.fields.length === 0,
+        payloads,
+      );
+    }),
+    // Now reduce them down to a single payload for each topic
+    R.map((payloads: Immutable<SubscribePayload[]> | undefined): Immutable<SubscribePayload> => {
+      const first = payloads?.[0];
+      if (first == undefined) {
+        // by definition, we can only get here iff there is at least one
```

Should probably throw an error

---

Pure functions ideally should not throw errors--it's a side effect, after all

---

I'd rather have this fail loudly than silently. If there's an invariant here it's better we know about it sooner rather than have it be silent

---

If the side effect is that we find a bug, it’s a good side effect 😀